plugins/chat/app/services/chat/update_message.rb

```diff
@@ -0,0 +1,150 @@
+# frozen_string_literal: true
+
+module Chat
+  # Service responsible for updating a message.
+  #
+  # @example
+  #  Chat::UpdateMessage.call(message_id: 2, guardian: guardian, message: "A new message")
+  #
+  class UpdateMessage
+    include Service::Base
+
+    # @!method call(message_id:, guardian:, message:, upload_ids:)
+    #   @param guardian [Guardian]
+    #   @param message_id [Integer]
+    #   @param message [String]
+    #   @param upload_ids [Array<Integer>] IDs of uploaded documents
+
+    contract
+    model :message
+    model :uploads, optional: true
+    policy :can_modify_channel_message
+    policy :can_modify_message
+
+    transaction do
+      step :modify_message
+      step :ensures_message_validity
+      step :save_message
+      step :save_revision
+      step :publish
+    end
+
+    class Contract
+      attribute :message_id, :string
+      attribute :message, :string
+      attribute :upload_ids, :array
+
+      validates :message_id, presence: true
+      validates :message, presence: true, if: -> { upload_ids.blank? }
+    end
+
+    private
+
+    def fetch_message(contract:, **)
+      ::Chat::Message
+        .strict_loading
+        .includes(
+          :chat_mentions,
+          :bookmarks,
+          :chat_webhook_event,
+          :uploads,
+          :revisions,
+          reactions: [:user],
+          thread: [:channel, last_message: [:user]],
+          chat_channel: [
+            :last_message,
+            :chat_channel_archive,
+            chatable: [:topic_only_relative_url, direct_message_users: [user: :user_option]],
+          ],
+          user: :user_status,
+        )
+        .find_by(id: contract.message_id)
+    end
+
+    def fetch_uploads(contract:, guardian:, **)
+      return [] if !SiteSetting.chat_allow_uploads
+      guardian.user.uploads.where(id: contract.upload_ids)
+    end
+
+    def can_modify_channel_message(guardian:, message:, **)
+      guardian.can_modify_channel_message?(message.chat_channel)
+    end
+
+    def can_modify_message(guardian:, message:, **)
+      guardian.can_edit_chat?(message)
+    end
+
+    def modify_message(contract:, message:, guardian:, **)
+      message.message = contract.message
+      message.last_editor_id = guardian.user.id
+
+      return if !SiteSetting.chat_allow_uploads
+
+      uploads = ::Upload.where(id: contract.upload_ids, user_id: guardian.user.id)
+      return if uploads.size != contract.upload_ids.to_a.size
+
+      new_upload_ids = uploads.map(&:id)
+      existing_upload_ids = message.upload_ids
+      difference = (existing_upload_ids + new_upload_ids) - (existing_upload_ids & new_upload_ids)
+      return if !difference.any?
+
+      message.upload_ids = new_upload_ids
+    end
+
+    def ensures_message_validity(message:, **)
+      message.valid?
+    end
```

Comment:
Im not sure this step is necessary, it's probably doing nothing?

We have `on_model_errors` which is supposed to handle this case